Content.Server/Shuttles/Systems/ArrivalsSystem.cs

```diff
@@ -241,6 +260,41 @@ private void OnPlayerSpawn(PlayerSpawningEvent ev)
         }
     }
 
+    private bool TryTeleportToMapSpawn(EntityUid player, EntityUid stationId, TransformComponent? transform = null)
+    {
+        if (!Resolve(player, ref transform))
+            return false;
+
+        var points = EntityQuery<SpawnPointComponent, TransformComponent>().ToList();
+        _random.Shuffle(points);
+
+        // Find a spawnpoint on the same map as the player is already on now.
+        foreach (var (spawnPoint, xform) in points)
+        {
+            if (spawnPoint.SpawnType != SpawnPointType.LateJoin || _station.GetOwningStation(spawnPoint.Owner, xform) != stationId)
+                continue;
```

Comment:
Use EntityQueryEnumerator and you can avoid .Owner.

Realistically we need a randomcomponent method on engine.

Reply:
... Yes, but we need it in list form and shuffled

Reply:
I solved it.